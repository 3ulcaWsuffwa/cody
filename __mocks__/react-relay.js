const ReactRelay = require("react-relay");

module.exports = {
  ...ReactRelay,
  // Mock createFragmentContainer HOC so it just returns the inner Component
  createFragmentContainer: Component => Component
};
