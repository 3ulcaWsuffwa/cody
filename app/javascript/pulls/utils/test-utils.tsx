import React from "react";
import { render } from "@testing-library/react";
import { IntlProvider } from "react-intl";
import { MemoryRouter } from "react-router";

// eslint-disable-next-line react/prop-types
const TestProvider = ({ children }) => {
  return (
    <IntlProvider locale="en" timeZone="UTC">
      <MemoryRouter>
        {children}
      </MemoryRouter>
    </IntlProvider>
  )
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
function customRender(ui, options?) {
  return render(ui, { wrapper: TestProvider, ...options });
}

export * from "@testing-library/react";
export { customRender as render };
