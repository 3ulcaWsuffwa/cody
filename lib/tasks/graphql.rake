namespace :graphql do
  desc "Export the GraphQL schema"
  task schema: :environment do
    File.write(
      Rails.root.join("schema.graphql"),
      GraphQL::Schema::Printer.print_schema(CodySchema)
    )
  end

  desc "Run Relay compiler"
  task relay: :schema do
    sh "bin/yarn relay"
  end
end
