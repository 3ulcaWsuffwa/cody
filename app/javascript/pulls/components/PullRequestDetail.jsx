// @flow

import React from "react";
import Reviewer from "./Reviewer";
import { createFragmentContainer, graphql } from "react-relay";
import { type PullRequestDetail_pullRequest } from "./__generated__/PullRequestDetail_pullRequest.graphql";

const PullRequestDetail = ({
  pullRequest
}: {
  pullRequest: PullRequestDetail_pullRequest
}) => {
  return (
    <section className="section">
      <div className="container box code">
        <div className="pa4 br2 ba b--light-gray mt4 bg-white">
          <h1 className="title code">
            {`${pullRequest.repository}#${pullRequest.number}`}
          </h1>
          <h2 className="subtitle code">
            {pullRequest.status}
          </h2>
        </div>
        <hr />
        <div>
          {pullRequest.reviewers.edges.map(edge => {
            return <Reviewer key={edge.node.id} reviewer={edge.node} />;
          })}
        </div>
      </div>
    </section>
  );
};

export default createFragmentContainer(
  PullRequestDetail,
  graphql`
    fragment PullRequestDetail_pullRequest on PullRequest {
      id
      repository
      number
      status
      reviewers {
        edges {
          node {
            ...Reviewer_reviewer
          }
        }
      }
    }
  `
);
