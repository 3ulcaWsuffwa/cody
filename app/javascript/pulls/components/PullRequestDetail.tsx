import React from "react";
import Reviewer from "./Reviewer";
import { createFragmentContainer, graphql } from "react-relay";
import { PullRequestDetail_pullRequest } from "./__generated__/PullRequestDetail_pullRequest.graphql";
import Container from "./Container";
import { GitPullRequest } from "react-feather";

function humanStatus(status: string): string {
  switch (status) {
    case "pending_review":
      return "Pending Review";
    case "approved":
      return "Approved";
    case "closed":
      return "Closed";
    default:
      return status;
  }
}

function PullRequestDetail({
  pullRequest,
}: {
  pullRequest: PullRequestDetail_pullRequest;
}): JSX.Element {
  return (
    <Container>
      <div className="bg-white shadow overflow-hidden sm:rounded-md">
        <div className="divide-y divide-gray-200">
          <div className="flex items-center py-4 px-4 sm:px-6">
            <div className="min-w-0 flex flex-1 items-center">
              <div className="flex-shrink-0">
                <GitPullRequest />
              </div>
              <div className="min-w-0 flex-1 px-4 md:grid md:grid-cols-2 md:gap-4">
                <div>
                  <h1 className="text-lg leading-6 font-medium text-gray-900">
                    {`${pullRequest.repository} #${pullRequest.number}`}
                  </h1>
                  <p className="mt-1 text-sm text-gray-500">
                    {humanStatus(pullRequest.status)}
                  </p>
                </div>
              </div>
            </div>
          </div>
          <div>
            <ul>
              {pullRequest.reviewers != null &&
              pullRequest.reviewers.edges != null
                ? pullRequest.reviewers.edges.map((edge) => {
                    if (edge != null && edge.node != null) {
                      // https://github.com/facebook/relay/issues/1918
                      return (
                        <Reviewer key={edge.node.id} reviewer={edge.node} />
                      );
                    }
                  })
                : null}
            </ul>
          </div>
        </div>
      </div>
    </Container>
  );
}

export default createFragmentContainer(PullRequestDetail, {
  pullRequest: graphql`
    fragment PullRequestDetail_pullRequest on PullRequest {
      id
      repository
      number
      status
      reviewers {
        edges {
          node {
            id
            ...Reviewer_reviewer
          }
        }
      }
    }
  `,
});
