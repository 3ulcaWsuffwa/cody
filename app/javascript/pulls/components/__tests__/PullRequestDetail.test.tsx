import React from "react";
import { render } from "test-utils";
import { createMockEnvironment, MockPayloadGenerator } from "relay-test-utils";
import { QueryRenderer, graphql, Environment } from "react-relay";
import PullRequestDetail from "../PullRequestDetail";
import { PullRequestDetailTestSnapshotQuery } from "./__generated__/PullRequestDetailTestSnapshotQuery.graphql";

const TestComponent = ({ environment }: { environment: Environment }) => (
  <QueryRenderer<PullRequestDetailTestSnapshotQuery>
    environment={environment}
    query={graphql`
      query PullRequestDetailTestSnapshotQuery @relay_test_operation {
        viewer {
          repository(owner: "test", name: "test") {
            pullRequest(number: "42") {
              ...PullRequestDetail_pullRequest
            }
          }
        }
      }
    `}
    variables={{}}
    render={({ error, props }) => {
      if (error) {
        return error.message;
      } else if (props) {
        return (
          <PullRequestDetail pullRequest={props.viewer.repository.pullRequest} />
        );
      } else {
        return "Loading";
      }
    }}
  />
);

let environment;
beforeEach(() => {
  environment = createMockEnvironment();
});

test("PullRequestDetail snapshot test", () => {
  const renderedComponent = render(<TestComponent environment={environment} />);

  environment.mock.resolveMostRecentOperation((operation) =>
    MockPayloadGenerator.generate(operation)
  );

  const fragment = renderedComponent.asFragment();
  expect(fragment).toMatchSnapshot();
});
