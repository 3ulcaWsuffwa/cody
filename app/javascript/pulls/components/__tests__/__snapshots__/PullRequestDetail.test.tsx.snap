// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PullRequestDetail snapshot test 1`] = `
<DocumentFragment>
  <div
    class="max-w-7xl mx-auto sm:px-6 lg:px-8 py-6"
  >
    <div
      class="bg-white shadow overflow-hidden sm:rounded-md"
    >
      <div
        class="divide-y divide-gray-200"
      >
        <div
          class="flex items-center py-4 px-4 sm:px-6"
        >
          <div
            class="min-w-0 flex flex-1 items-center"
          >
            <div
              class="flex-shrink-0"
            >
              <svg
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <circle
                  cx="18"
                  cy="18"
                  r="3"
                />
                <circle
                  cx="6"
                  cy="6"
                  r="3"
                />
                <path
                  d="M13 6h3a2 2 0 0 1 2 2v7"
                />
                <line
                  x1="6"
                  x2="6"
                  y1="9"
                  y2="21"
                />
              </svg>
            </div>
            <div
              class="min-w-0 flex-1 px-4 md:grid md:grid-cols-2 md:gap-4"
            >
              <div>
                <h1
                  class="text-lg leading-6 font-medium text-gray-900"
                >
                  &lt;mock-value-for-field-"repository"&gt; #&lt;mock-value-for-field-"number"&gt;
                </h1>
                <p
                  class="mt-1 text-sm text-gray-500"
                >
                  &lt;mock-value-for-field-"status"&gt;
                </p>
              </div>
            </div>
          </div>
        </div>
        <div>
          <ul>
            <li>
              <div
                class="flex items-center px-4 py-4 sm:px-6"
              >
                <div
                  class="min-w-0 flex flex-1 items-center"
                >
                  <div
                    class="flex-shrink-0"
                  >
                    <span
                      class="sr-only"
                    >
                      Approved
                    </span>
                    <svg
                      class="flex-shrink-0 mr-1.5 h-5 w-5 text-green-500"
                      fill="none"
                      height="24"
                      stroke="currentColor"
                      stroke-linecap="round"
                      stroke-linejoin="round"
                      stroke-width="2"
                      viewBox="0 0 24 24"
                      width="24"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M22 11.08V12a10 10 0 1 1-5.93-9.14"
                      />
                      <polyline
                        points="22 4 12 14.01 9 11.01"
                      />
                    </svg>
                  </div>
                  <div
                    class="min-w-0 flex-1 px-4 md:grid md:grid-cols-2 md:gap-4"
                  >
                    <div>
                      <p
                        class="text-sm font-medium truncate"
                        data-testid="reviewer-login"
                      >
                        &lt;mock-value-for-field-"login"&gt;
                      </p>
                      <p
                        class="text-sm truncate text-gray-500"
                        data-testid="review-rule-name"
                      >
                        &lt;mock-value-for-field-"name"&gt;
                      </p>
                    </div>
                  </div>
                </div>
              </div>
            </li>
          </ul>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
